src/graph/Graph.cpp

```diff
@@ -70,15 +70,16 @@ void Graph::run()
 		rt->setFields(fieldsToCompute);
 	}
 
-	for (auto&& current : nodesInExecOrder) {
-		RGL_DEBUG("Validating node: {}", *current);
-		current->onInputChange();
+	for (auto&& node : nodesInExecOrder) {
+		if (!node->canExecute()) {
+			auto msg = fmt::format("cannot execute graph - Node {} is not ready to run (missing input?)", node->getName());
+			throw InvalidPipeline(msg);
+		}
 	}
-	RGL_DEBUG("Node validation completed");  // This also logs the time diff for the last one.
 
 	for (auto&& node : nodesInExecOrder) {
-		RGL_DEBUG("Scheduling node: {}", *node);
-		node->schedule(nullptr);
+		RGL_DEBUG("Executing node: {}", *node);
```

Comment:
I get the idea, however this may suggest, that this line relly execute the workload, not only push order to Q.